File: test-fixtures/languages/scala/micro/Gateway.scala
package micro

import scala.annotation.StaticAnnotation
import javax.servlet.http.{HttpServletRequest, HttpServletResponse}

class Audited extends StaticAnnotation

class Gateway {
  private val userService = new UserService()

  @Audited
  def handleRequest(req: HttpServletRequest, res: HttpServletResponse): Unit = {
    val token = req.getParameter("token")    // source: user input
    val action = req.getParameter("action")  // source: user input

    val user = userService.getUser(token)       // flows to SQL injection
    val result = userService.updateUser(token, action) // flows to command injection

    res.getWriter.write(s"""{"user": "${user.id}", "result": "$result"}""")
  }
}

File: test-fixtures/languages/c/micro/gateway.c
#include "gateway.h"
#include "user_service.h"
#include <stdio.h>
#include <stdlib.h>
#include <string.h>

void handle_request(const char *query_string) {
    char token[128];
    char action[128];
    sscanf(query_string, "token=%127[^&]&action=%127s", token, action); /* source: user input */

    UserInfo *user = get_user(token);         /* flows to SQL injection */
    char *result = update_user(token, action); /* flows to command injection */

    printf("{\"user\": \"%s\", \"result\": \"%s\"}\n",
           user ? user->id : "null", result ? result : "null");
    free(user);
    free(result);
}
